src/datanode/src/region_server.rs

```diff
@@ -306,48 +348,109 @@ impl RegionServerInner {
             | RegionRequest::Truncate(_) => RegionChange::None,
         };
 
-        let engine = match &region_change {
-            RegionChange::Register(engine_type) => self
-                .engines
-                .read()
-                .unwrap()
-                .get(engine_type)
-                .with_context(|| RegionEngineNotFoundSnafu { name: engine_type })?
-                .clone(),
-            RegionChange::None | RegionChange::Deregisters => self
-                .region_map
-                .get(&region_id)
-                .with_context(|| RegionNotFoundSnafu { region_id })?
-                .clone(),
+        let region_status = self.region_map.get(&region_id);
+
+        let engine = match region_change {
+            RegionChange::Register(ref engine_type) => match region_status {
+                Some(status) => {
+                    if status.is_registering() {
+                        return Ok(0);
+                    } else {
+                        status.clone().into_engine()
+                    }
+                }
+                _ => self
+                    .engines
+                    .read()
+                    .unwrap()
+                    .get(engine_type)
+                    .with_context(|| RegionEngineNotFoundSnafu { name: engine_type })?
+                    .clone(),
+            },
+            RegionChange::Deregisters => match region_status {
+                Some(status) => {
+                    if status.is_deregistering() {
+                        return Ok(0);
+                    } else {
+                        status.clone().into_engine()
+                    }
+                }
+                None => return Ok(0),
+            },
+            RegionChange::None => match region_status {
+                Some(status) => match status.clone() {
+                    RegionEngineWithStatus::Registering(_) => {
+                        return error::RegionNotReadySnafu { region_id }.fail()
+                    }
+                    RegionEngineWithStatus::Deregistering(_) => {
+                        return error::RegionNotFoundSnafu { region_id }.fail()
+                    }
+                    RegionEngineWithStatus::Ready(engine) => engine,
+                },
+                None => return error::RegionNotFoundSnafu { region_id }.fail(),
+            },
```

Comment:
```suggestion
            RegionChange::None => match region_status {
                Some(RegionEngineWithStatus::Ready(engine)) => engine.clone()
                _ => return error::RegionNotFoundSnafu { region_id }.fail(),
            },
```

Reply:
Unfortunately, we can't do that; it needs to let the caller know the region is not ready.